File: repo/frontend/src/types/order.ts
export type OrderStatus =
  | 'pending_payment'
  | 'pending_fulfillment'
  | 'pending_receipt'
  | 'completed'
  | 'canceled'
  | 'refund_in_progress'
  | 'refunded'

export type PricingMode = 'fixed' | 'bargaining'

// Shape mirrors backend `schemas/order.py::ServiceItemRead`. Prior FE version
// used `base_price`/`is_active`/`pricing_mode: string` which the backend does
// not return — the real catalog uses `item_code`, `fixed_price` and a
// `bargaining_enabled` flag alongside `pricing_mode`.
export interface ServiceItem {
  id: string
  item_code: string
  name: string
  description: string | null
  pricing_mode: PricingMode
  fixed_price: string | null
  is_capacity_limited: boolean
  bargaining_enabled: boolean
  available_slots: number | null
}

// Shape mirrors backend `schemas/order.py::OrderEventRead`. Prior FE version
// used `order_id`/`prev_state`/`created_at` — backend emits `sequence_number`,
// `previous_state`, `occurred_at` instead and does not echo `order_id`.
export interface OrderEvent {
  id: string
  sequence_number: number
  event_type: string
  previous_state: OrderStatus | null
  new_state: OrderStatus
  actor_id: string | null
  actor_role: string | null
  notes: string | null
  occurred_at: string
}

export interface PaymentRecord {
  id: string
  order_id: string
  amount: string
  payment_method: string
  reference_number: string
  notes: string | null
  confirmed_by: string | null
  confirmed_at: string | null
  created_at: string
}

export interface BargainingOffer {
  id: string
  thread_id: string
  offer_number: number
  amount: string
  submitted_by: string
  outcome: 'pending' | 'accepted' | 'expired'
  created_at: string
}

export type BargainingThreadStatus =
  | 'open'
  | 'accepted'
  | 'countered'
  | 'counter_accepted'
  | 'expired'

export interface BargainingThread {
  id: string
  order_id: string
  status: BargainingThreadStatus
  window_starts_at: string
  window_expires_at: string
  offers: BargainingOffer[]
  counter_amount: string | null
  counter_count: number
  counter_by: string | null
  counter_at: string | null
  resolved_offer_id: string | null
  resolved_at: string | null
  created_at: string
}

export interface Voucher {
  id: string
  order_id: string
  voucher_code: string
  issued_by: string
  notes: string | null
  created_at: string
}

export interface FulfillmentMilestone {
  id: string
  order_id: string
  milestone_type: string
  description: string | null
  recorded_by: string
  occurred_at: string
  created_at: string
}

export interface RefundRecord {
  id: string
  order_id: string
  amount: string
  initiated_by: string
  initiated_at: string
  processed_by: string | null
  processed_at: string | null
  rollback_applied: boolean
  reason: string | null
}

export type AfterSalesStatus = 'open' | 'resolved'

export interface AfterSalesRequest {
  id: string
  order_id: string
  requested_by: string
  request_type: string
  description: string
  status: AfterSalesStatus
  window_expires_at: string
  resolved_by: string | null
  resolved_at: string | null
  resolution_notes: string | null
  created_at: string
}

// Shape mirrors backend `schemas/order.py::OrderRead`. The backend response
// does NOT include `item_name`, `base_price`, `idempotency_key`,
// `payment_record`, or `bargaining_thread` — views should resolve the service
// item via `serviceApi.getServiceItem(item_id)` (or the service-item store)
// for display pricing/naming, and fetch the bargaining thread / payment
// proofs via their own endpoints.
export interface Order {
  id: string
  candidate_id: string
  item_id: string
  status: OrderStatus
  pricing_mode: PricingMode
  agreed_price: string | null
  auto_cancel_at: string | null
  completed_at: string | null
  canceled_at: string | null
  cancellation_reason: string | null
  created_at: string
  updated_at: string
  events: OrderEvent[]
}

export interface OrderCreate {
  item_id: string
  pricing_mode?: PricingMode
  notes?: string | null
}

export interface PaymentProofSubmit {
  payment_method: string
  reference_number: string
  amount: string
  notes?: string | null
}

export interface PaymentConfirmPayload {
  amount: string
  payment_method: string
  reference_number?: string | null
  notes?: string | null
}

export interface AfterSalesRequestCreate {
  request_type: string
  description: string
}

export interface VoucherCreate {
  notes?: string | null
}

export interface MilestoneCreate {
  milestone_type: string
  description?: string | null
  occurred_at?: string
}

export interface RefundCreate {
  amount: string
  reason?: string | null
}

File: repo/backend/src/schemas/order.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from ..domain.enums import OrderStatus, PricingMode


class ServiceItemRead(BaseModel):
    id: uuid.UUID
    item_code: str
    name: str
    description: str | None
    pricing_mode: PricingMode
    fixed_price: Decimal | None
    is_capacity_limited: bool
    bargaining_enabled: bool
    available_slots: int | None = None


class OrderCreate(BaseModel):
    item_id: uuid.UUID
    pricing_mode: PricingMode
    idempotency_key: str | None = Field(None, max_length=64)


class OrderEventRead(BaseModel):
    id: uuid.UUID
    sequence_number: int
    event_type: str
    previous_state: OrderStatus | None
    new_state: OrderStatus
    actor_id: uuid.UUID | None
    actor_role: str | None
    notes: str | None
    occurred_at: datetime


class OrderRead(BaseModel):
    id: uuid.UUID
    candidate_id: uuid.UUID
    item_id: uuid.UUID
    status: OrderStatus
    pricing_mode: PricingMode
    agreed_price: Decimal | None
    auto_cancel_at: datetime | None
    completed_at: datetime | None
    canceled_at: datetime | None
    cancellation_reason: str | None
    created_at: datetime
    updated_at: datetime
    events: list[OrderEventRead] = []


class PaymentConfirmRequest(BaseModel):
    amount: Decimal = Field(..., gt=0)
    payment_method: str = Field(..., max_length=50)
    reference_number: str | None = Field(None, max_length=200)
    notes: str | None = None


class VoucherRead(BaseModel):
    id: uuid.UUID
    order_id: uuid.UUID
    voucher_code: str
    issued_by: uuid.UUID
    issued_at: datetime
    notes: str | None


class MilestoneCreate(BaseModel):
    milestone_type: str = Field(..., max_length=100)
    description: str | None = None


class MilestoneRead(BaseModel):
    id: uuid.UUID
    order_id: uuid.UUID
    milestone_type: str
    description: str | None
    recorded_by: uuid.UUID
    occurred_at: datetime


class RefundCreate(BaseModel):
    amount: Decimal = Field(..., gt=0)
    reason: str = Field(..., min_length=1, max_length=2000)


class RefundRead(BaseModel):
    id: uuid.UUID
    order_id: uuid.UUID
    amount: Decimal
    initiated_by: uuid.UUID
    initiated_at: datetime
    processed_by: uuid.UUID | None
    processed_at: datetime | None
    reason: str
    rollback_applied: bool


class PaymentProofSubmit(BaseModel):
    amount: Decimal = Field(..., gt=0)
    payment_method: str = Field(..., max_length=50)
    reference_number: str | None = Field(None, max_length=200)
    notes: str | None = None


class VoucherCreate(BaseModel):
    notes: str | None = Field(None, max_length=1000)


class OrderTransitionRequest(BaseModel):
    notes: str | None = Field(None, max_length=500)


class AfterSalesRequestCreate(BaseModel):
    request_type: str = Field(..., max_length=100)
    description: str = Field(..., min_length=1, max_length=5000)


class AfterSalesRequestRead(BaseModel):
    id: uuid.UUID
    order_id: uuid.UUID
    requested_by: uuid.UUID
    request_type: str
    description: str
    status: str
    window_expires_at: datetime
    resolved_by: uuid.UUID | None
    resolved_at: datetime | None
    resolution_notes: str | None
    created_at: datetime


class AfterSalesResolveRequest(BaseModel):
    resolution_notes: str = Field(..., min_length=1, max_length=5000)
